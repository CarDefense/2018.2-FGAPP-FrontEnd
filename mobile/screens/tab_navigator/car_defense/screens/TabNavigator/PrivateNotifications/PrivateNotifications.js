import { TextField } from 'react-native-material-textfield';
import { Constants, ImagePicker, Permissions } from 'expo';
import React, { Component } from 'react';
import {
  ActivityIndicator,
  View,
  Text,
  StyleSheet,
  ScrollView,
  TextInput,
  TouchableOpacity,
  Button,
  Alert,
  Image
} from 'react-native';

var tk
async function register() {
  const { status } = await Expo.Permissions.askAsync(
    Expo.Permissions.NOTIFICATIONS
  );
  if (status != 'granted') {
    alert('You need to enable permissions in settings');
    return;
  }

  const value = await Expo.Notifications.getExpoPushTokenAsync();
  tk = value;
  console.log(status, value);
}

export default class PrivateNotifications extends Component {
  componentWillMount() {
    register();
    this.listener = Expo.Notifications.addListener(this.listen);
  }
  componentWillUnmount() {
    this.listener && Expo.Notifications.addListener(this.listen);
  }

  listen = ({ origin, data }) => {
    console.log('cool data', origin, data);
  }

  constructor(props) {
    super(props);

    this.plateRef = this.updateRef.bind(this, 'plate');
    this.messageRef = this.updateRef.bind(this, 'message');
    this.onFocus = this.onFocus.bind(this);

    this.state = {
      title: 'Notificação',
      plate: '',
      message: '',
      image: null,
      uploading: false,
    }
  }
   
  onFocus() {
    let { errors = {} } = this.state;

    for (let text in errors) {
      let ref = this[text];

      if (ref && ref.isFocused()) {
        delete errors[text];
      }
    }

    this.setState({ errors });
  }

  onChangeText(words) {
    ['plate']
    .map((text) => ({ text, ref: this[text] }))
    .forEach(({ text, ref }) => {
      if (ref && ref.isFocused()) {
        this.setState({ [text]: words });
      }
    });
    ['message']
    .map((text) => ({ text, ref: this[text] }))
    .forEach(({ text, ref }) => {
      if (ref && ref.isFocused()) {
        this.setState({ [text]: words });
      }
    });
  }

  // onSubmitFirstName() {
  //   this.lastname.focus();
  // }

  onPressButton = () => {
    let errors = {};
    const url = 'http://192.168.1.4:8002/send_push_message/' //function send_push_message url
    let errorPlate = false;
    let errorMessage = false;

    ['plate']
    .forEach((text) => {
      let value = this[text].value();

      if (!value) {
        errors[text] = 'Campo obrigatório.';
        errorPlate = true;
      } else {
        if ('plate' === text && value.length < 8) {
          errors[text] = 'Placa inválida. Ex.:AAA-0000';
          errorPlate = true;
        }
      }
    });

    ['message']
    .forEach((text) => {
      let value = this[text].value();

      if (!value) {
        errors[text] = 'Campo obrigatório.';
        errorMessage = true;
      } else {
        if ('message' === text && value.length < 5) {
          errors[text] = 'Forneça mais detalhes do ocorrido.';
          errorMessage = true;
        }
      }
    });

    if (errorPlate == false && errorMessage == false) {
      let notification = JSON.stringify({
        sender_id: tk,
        title: this.state.title,
        plate: this.state.plate,
        message: this.state.message,
        image: this.state.image
      })

      console.log(notification);

      fetch(url, {
        method: 'POST',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: notification
      }).then(response => { return response.json() }
      ).then(jsonResponse => {
        console.log(jsonResponse);
        Alert.alert("Notificação enviada!")
      }      
      ).catch(error => {
        console.log(error)
        Alert.alert("Placa não existe!")
      })
    }
    this.setState({ errors });
  }

  updateRef(text, ref) {
    this[text] = ref;
  }

  render() {
    let { image } = this.state;
    let { errors = {}, ...data} = this.state;
    let {plate = 'text'} = data;
    let {message = 'text'} = data;

    return (
      <ScrollView>
        <View style={styles.container}>
          <Text style={styles.header}>Notificações</Text>
          <TextField
            ref={this.plateRef}
            value={data.plate}
            autoCorrect={false}
            enablesReturnKeyAutomatically={true}
            onFocus={this.onFocus}
            onChangeText={ (plate) => this.setState({ plate }) }
            // onChangeText={this.onChangeText}
            // onSubmitEditing={this.onSubmitPLate}
            returnKeyType='next'
            label='Placa'
            tintColor = "#760f9f"
            underlineColorAndroid="transparent"
            maxLength={8}
            autoCapitalize="characters"
            error={errors.plate}
          />
          <TextField
            ref={this.messageRef}
            value={data.message}
            autoCorrect={false}
            enablesReturnKeyAutomatically={true}
            onFocus={this.onFocus}
            onChangeText={ (message) => this.setState({ message }) }
            // onChangeText={this.onChangeText}
            // onSubmitEditing={this.onSubmitMessage}
            returnKeyType='next'
            label='Descrição'
            tintColor = "#760f9f"
            error={errors.message}
          />
          <Text style={styles.text}>Adicionar imagem</Text>
          <View style={styles.alternativeLayoutButtonContainer}>
            <TouchableOpacity
              style={styles.button}
              color="#760f9f"
              onPress={this._takePhoto}
              containerViewStyle={{ width: '40%' }}
            >
              <Text style={{ color: 'white' }} >Camera</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.button2}
              color="#540b71"
              onPress={this._pickImage}
              containerViewStyle={{ width: '40%' }}
            >
              <Text style={{ color: 'white' }} >Galeria</Text>
            </TouchableOpacity>
          </View>
          {this._maybeRenderImage()}
          {this._maybeRenderUploadingOverlay()}
          <View style={styles.container1}>
            <TouchableOpacity
              style={styles.button3}
              color="#760f9f"
              onPress={this.onPressButton}
              containerViewStyle={{ width: '40%' }}
            >
              <Text style={{ color: 'white' }} >Enviar</Text>
            </TouchableOpacity>
          </View>
        </View>
      </ScrollView>
    );
  }

  _maybeRenderUploadingOverlay = () => { //done
    if (this.state.uploading) {
      return (
        <View
          style={styles.maybeRenderUploading}>
          <ActivityIndicator color="#313869" size="large" />
        </View>
      );
    }
  };

  _maybeRenderImage = () => { //done
    let {
      image
    } = this.state;

    if (!image) {
      return;
    }

    // onPress={this._copyToClipboard}
    // onLongPress={this._share}

    return (
      <View
        style={styles.maybeRenderContainer}>
        <View
          style={styles.maybeRenderImageContainer}>
          <Image source={{ uri: image }} style={styles.maybeRenderImage} />
        </View>
      </View>
    );
  };

  _share = () => {
    Share.share({
      message: this.state.image,
      title: 'Check out this photo',
      url: this.state.image,
    });
  };

  _copyToClipboard = () => {
    Clipboard.setString(this.state.image);
    alert('Copied image URL to clipboard');
  };

  _takePhoto = async () => { //Done
    const {
      status: cameraPerm
    } = await Permissions.askAsync(Permissions.CAMERA);

    const {
      status: cameraRollPerm
    } = await Permissions.askAsync(Permissions.CAMERA_ROLL);

  // only if user allows permission to camera AND camera roll
    if (cameraPerm === 'granted' && cameraRollPerm === 'granted') {
      let pickerResult = await ImagePicker.launchCameraAsync({
        allowsEditing: true,
      });

      this._handleImagePicked(pickerResult);
    }
  };

  _pickImage = async () => {
    const {
      status: cameraRollPerm
    } = await Permissions.askAsync(Permissions.CAMERA_ROLL);

  // only if user allows permission to camera roll
    if (cameraRollPerm === 'granted') {
      let pickerResult = await ImagePicker.launchImageLibraryAsync({
        allowsEditing: true,

      });

      this._handleImagePicked(pickerResult);
    }
  };

  _handleImagePicked = async pickerResult => {
    let uploadResponse, uploadResult;

    try {
      this.setState({
        uploading: true
      });

      if (!pickerResult.cancelled) {
        uploadResponse = await uploadImageAsync(pickerResult.uri);
        uploadResult = await uploadResponse.json();

        console.log(uploadResult.image)

        this.setState({
          image: uploadResult.image
        });
      }
    } catch (e) {
      console.log({ uploadResponse });
      console.log({ uploadResult });
      console.log({ e });
      alert('Upload failed, sorry :(');
    } finally {
      this.setState({
      uploading: false
      });
    }
  };
}

  async function uploadImageAsync(uri) {
    let apiUrl = 'http://192.168.1.4:8002/notificationsimage/';

    // Note:
    // Uncomment this if you want to experiment with local server
    //
    // if (Constants.isDevice) {
    //   apiUrl = `https://your-ngrok-subdomain.ngrok.io/upload`;
    // } else {
    //   apiUrl = `http://localhost:3000/upload`
    // }

    let uriParts = uri.split('.');
    let fileType = uriParts[uriParts.length - 1];

    let formData = new FormData();
    formData.append('image', {
      uri,
      name: `photo.${fileType}`,
      type: `image/${fileType}`,
    });

    console.log(uri)

    let options = {
      method: 'POST',
      body: formData,
      headers: {
        Accept: 'application/json',
        'Content-Type': 'multipart/form-data',
      },
    };
    return fetch(apiUrl, options);
  }

const styles = StyleSheet.create({
  container: {
    paddingLeft: 16,
    paddingRight: 16
  },
  text: {
    paddingTop: 16,
    color: "#760f9f",
    fontWeight: 'bold',
  },
  alternativeLayoutButtonContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingTop: 4,
    paddingRight: 4,
  },
  button: {
    backgroundColor: "#760f9f",
    borderRadius: 15,
    height: 40,
    width: 190,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  button2: {
    backgroundColor: "#540b71",
    borderRadius: 15,
    height: 40,
    width: 120,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  button3: {
    backgroundColor: "#760f9f",
    borderRadius: 15,
    height: 40,
    width: 120,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center'
  },
  container1: {
    paddingTop: 20,
    flexDirection: 'row',
    justifyContent: 'center'
  },
  header: {
    color: "#760f9f",
    textAlign: 'center',
    fontWeight: '200',
    fontSize: 50,
    marginTop: 25
  },
  maybeRenderUploading: {
    paddingTop: 8,
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.0)',
    justifyContent: 'center',
  },
  maybeRenderContainer: {
    width: '100%',
    height: '40%',
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 3,
    elevation: 2,
    marginTop: 32,
    shadowColor: 'rgba(0,0,0,1)',
    shadowOpacity: 0.2,
    shadowOffset: {
      height: 4,
      width: 4,
    },
    shadowRadius: 5
  },
  maybeRenderImageContainer: {
    borderTopLeftRadius: 3,
    borderTopRightRadius: 3,
    overflow: 'hidden',
  },
  maybeRenderImage: {
    height: 250,
    width: 250,
  }
});
