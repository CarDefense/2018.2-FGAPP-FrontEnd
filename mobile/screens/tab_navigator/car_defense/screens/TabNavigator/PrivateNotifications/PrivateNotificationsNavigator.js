import React, { Component } from "react";
import {StackNavigator} from 'react-navigation'

import PrivateNotifications from './PrivateNotifications'


export default class PrivateNotificationsNavigator extends Component {

    render() {
        return (
            <PrivateNotificationsStackNavigator />
        );
    }
}

const PrivateNotificationsStackNavigator = new StackNavigator({

    PrivateNotifications: {
        screen: PrivateNotifications,
        navigationOptions: ({ navigation }) => ({
            header: null,
        }),
    },
})
