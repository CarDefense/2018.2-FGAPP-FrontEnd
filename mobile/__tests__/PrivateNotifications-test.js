import PrivateNotifications from '../screens/tab_navigator/car_defense/screens/TabNavigator/PrivateNotifications/PrivateNotifications'
import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

test('PrivateNotifications snapShot', () => {
    const snap = renderer.create(
        <PrivateNotifications/>
    ).toJSON()
    expect(snap).toMatchSnapshot()
})