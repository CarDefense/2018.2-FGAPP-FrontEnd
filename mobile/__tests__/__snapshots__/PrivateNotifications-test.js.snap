// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateNotifications snapShot 1`] = `
<RCTScrollView>
  <View>
    <View
      style={
        Object {
          "paddingLeft": 16,
          "paddingRight": 16,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#760f9f",
            "fontSize": 50,
            "fontWeight": "200",
            "marginTop": 25,
            "textAlign": "center",
          }
        }
      >
        Notificação
      </Text>
      <View
        onResponderRelease={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="auto"
        style={undefined}
      >
        <View
          collapsable={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "rgba(0, 0, 0, 0.38)",
              "borderBottomWidth": 0.5,
              "height": 64,
              "paddingBottom": 8,
              "paddingTop": 32,
            }
          }
        >
          <View
            collapsable={undefined}
            style={
              Object {
                "position": "absolute",
                "top": 36,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.38)",
                  "fontSize": 16,
                }
              }
            >
              Placa
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="characters"
              autoCorrect={false}
              disableFullscreenUI={true}
              editable={true}
              enablesReturnKeyAutomatically={true}
              error={undefined}
              maxLength={8}
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onContentSizeChange={[Function]}
              onFocus={[Function]}
              returnKeyType="next"
              selectionColor="#760f9f"
              style={
                Array [
                  Object {
                    "flex": 1,
                    "margin": 0,
                    "padding": 0,
                    "top": 2,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, .87)",
                    "fontSize": 16,
                    "height": 24,
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "flexDirection": "row",
              "height": 8,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgb(213, 0, 0)",
                    "fontSize": 0,
                    "opacity": 0,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0, 0, 0, .38)",
                    "fontSize": 12,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onResponderRelease={[Function]}
        onStartShouldSetResponder={[Function]}
        pointerEvents="auto"
        style={undefined}
      >
        <View
          collapsable={undefined}
          style={
            Object {
              "backgroundColor": "transparent",
              "borderBottomColor": "rgba(0, 0, 0, 0.38)",
              "borderBottomWidth": 0.5,
              "height": 64,
              "paddingBottom": 8,
              "paddingTop": 32,
            }
          }
        >
          <View
            collapsable={undefined}
            style={
              Object {
                "position": "absolute",
                "top": 36,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              collapsable={undefined}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                Object {
                  "color": "rgba(0, 0, 0, 0.38)",
                  "fontSize": 16,
                }
              }
            >
              Descrição
            </Text>
          </View>
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              autoCapitalize="sentences"
              autoCorrect={false}
              disableFullscreenUI={true}
              editable={true}
              enablesReturnKeyAutomatically={true}
              error={undefined}
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onContentSizeChange={[Function]}
              onFocus={[Function]}
              returnKeyType="next"
              selectionColor="#760f9f"
              style={
                Array [
                  Object {
                    "flex": 1,
                    "margin": 0,
                    "padding": 0,
                    "top": 2,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, .87)",
                    "fontSize": 16,
                    "height": 24,
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          collapsable={undefined}
          style={
            Object {
              "flexDirection": "row",
              "height": 8,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgb(213, 0, 0)",
                    "fontSize": 0,
                    "opacity": 0,
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "bottom": 0,
                  "left": 0,
                  "paddingVertical": 4,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                collapsable={undefined}
                ellipsizeMode="tail"
                numberOfLines={1}
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "rgba(0, 0, 0, .38)",
                    "fontSize": 12,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#760f9f",
            "fontWeight": "bold",
            "paddingTop": 16,
          }
        }
      >
        Adicionar imagem
      </Text>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingRight": 4,
            "paddingTop": 4,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#760f9f",
              "borderRadius": 15,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "width": 190,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
              }
            }
          >
            Camera
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#540b71",
              "borderRadius": 15,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "width": 120,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
              }
            }
          >
            Galeria
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "paddingTop": 20,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          collapsable={undefined}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#760f9f",
              "borderRadius": 15,
              "flexDirection": "row",
              "height": 40,
              "justifyContent": "center",
              "opacity": 1,
              "width": 120,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "white",
              }
            }
          >
            Enviar
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
